File: src/main/java/Animals/Animal.java
package Animals;


import java.time.LocalDate;
import java.util.UUID;

public class Animal {
    private String Name;
    private Gender Gender;
    private Reservor ReservedBy;

    public Animal(String name, Gender gender) {
        this.Name = name;
        this.Gender = gender;
    }

    private void setName(String name) {
        Name = name;
    }

    public String getName() {
        return Name;
    }

    private void setGender(Gender gender) {
        Gender = gender;
    }

    public Gender getGender() {
        return Gender;
    }

    public void setReserver(Reservor reservor) {
        ReservedBy = reservor;
    }

    public Reservor getReserver() {
        return ReservedBy;
    }

    public boolean Reserve(String reservedBy) {
        if (this.ReservedBy == (null)) {
            this.ReservedBy = new Reservor(reservedBy, LocalDate.now());
            return true;
        }
        return false;
    }

    @Override
    public String toString() {
        String reserved = "not reserved";
        if (this.ReservedBy != (null)) {
            reserved = String.format("reserved by %s", ReservedBy.getName());
        }
        return String.format("%s, %s, %s", this.Name, this.Gender, reserved);
    }
}

File: src/main/kotlin/animal/Animal.kt
package main.kotlin.animal

import main.kotlin.ApplicationProcess
import main.kotlin.webshop.ISellable
import java.time.LocalDateTime

abstract class Animal(override val name: String, val gender: Gender, override val price: Double) :
    ISellable {
    var reservedBy: IReservor? = null
        private set

    fun reserve(reservedBy: String, reservedAt: LocalDateTime = LocalDateTime.now()): Boolean {
        if (this.reservedBy!=null) return false
        this.reservedBy = Reservor(reservedBy, reservedAt)
        return true
    }

    override fun toString(): String {
        val reserved: String = reservedBy?.let { "reserved by ${reservedBy!!.name}" } ?: "not reserved"
        return "$name, $gender, $reserved, $price"
    }

    //TODO: Threadsafe solution?
    init {
        ApplicationProcess.webshop.addProduct(this)
    }
}